// @mochiapp/js/dist/contents/video/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaylistEpisodeServerSkipType = exports.PlaylistEpisodeServerSubtitleFormat = exports.PlaylistEpisodeServerFormatType = exports.PlaylistEpisodeServerQualityType = void 0;
var PlaylistEpisodeServerQualityType;
(function (PlaylistEpisodeServerQualityType) {
    PlaylistEpisodeServerQualityType[PlaylistEpisodeServerQualityType["auto"] = 0] = "auto";
    PlaylistEpisodeServerQualityType[PlaylistEpisodeServerQualityType["q360p"] = 1] = "q360p";
    PlaylistEpisodeServerQualityType[PlaylistEpisodeServerQualityType["q480p"] = 2] = "q480p";
    PlaylistEpisodeServerQualityType[PlaylistEpisodeServerQualityType["q720p"] = 3] = "q720p";
    PlaylistEpisodeServerQualityType[PlaylistEpisodeServerQualityType["q1080p"] = 4] = "q1080p";
})(PlaylistEpisodeServerQualityType || (exports.PlaylistEpisodeServerQualityType = PlaylistEpisodeServerQualityType = {}));
var PlaylistEpisodeServerFormatType;
(function (PlaylistEpisodeServerFormatType) {
    PlaylistEpisodeServerFormatType[PlaylistEpisodeServerFormatType["hsl"] = 0] = "hsl";
    PlaylistEpisodeServerFormatType[PlaylistEpisodeServerFormatType["dash"] = 1] = "dash";
})(PlaylistEpisodeServerFormatType || (exports.PlaylistEpisodeServerFormatType = PlaylistEpisodeServerFormatType = {}));
var PlaylistEpisodeServerSubtitleFormat;
(function (PlaylistEpisodeServerSubtitleFormat) {
    PlaylistEpisodeServerSubtitleFormat[PlaylistEpisodeServerSubtitleFormat["vtt"] = 0] = "vtt";
    PlaylistEpisodeServerSubtitleFormat[PlaylistEpisodeServerSubtitleFormat["ass"] = 1] = "ass";
    PlaylistEpisodeServerSubtitleFormat[PlaylistEpisodeServerSubtitleFormat["srt"] = 2] = "srt";
})(PlaylistEpisodeServerSubtitleFormat || (exports.PlaylistEpisodeServerSubtitleFormat = PlaylistEpisodeServerSubtitleFormat = {}));
var PlaylistEpisodeServerSkipType;
(function (PlaylistEpisodeServerSkipType) {
    PlaylistEpisodeServerSkipType[PlaylistEpisodeServerSkipType["opening"] = 0] = "opening";
    PlaylistEpisodeServerSkipType[PlaylistEpisodeServerSkipType["ending"] = 1] = "ending";
    PlaylistEpisodeServerSkipType[PlaylistEpisodeServerSkipType["recap"] = 2] = "recap";
})(PlaylistEpisodeServerSkipType || (exports.PlaylistEpisodeServerSkipType = PlaylistEpisodeServerSkipType = {}));
