// @mochiapp/js/dist/interfaces/source/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaylistPreviewType = exports.PlaylistType = exports.PlaylistStatus = exports.DiscoverListingOrientationType = exports.DiscoverListingType = void 0;
var DiscoverListingType;
(function (DiscoverListingType) {
    DiscoverListingType[DiscoverListingType["default"] = 0] = "default";
    DiscoverListingType[DiscoverListingType["rank"] = 1] = "rank";
    DiscoverListingType[DiscoverListingType["featured"] = 2] = "featured";
})(DiscoverListingType || (exports.DiscoverListingType = DiscoverListingType = {}));
var DiscoverListingOrientationType;
(function (DiscoverListingOrientationType) {
    DiscoverListingOrientationType[DiscoverListingOrientationType["portrait"] = 0] = "portrait";
    DiscoverListingOrientationType[DiscoverListingOrientationType["landscape"] = 1] = "landscape";
})(DiscoverListingOrientationType || (exports.DiscoverListingOrientationType = DiscoverListingOrientationType = {}));
var PlaylistStatus;
(function (PlaylistStatus) {
    PlaylistStatus[PlaylistStatus["unknown"] = 0] = "unknown";
    PlaylistStatus[PlaylistStatus["upcoming"] = 1] = "upcoming";
    PlaylistStatus[PlaylistStatus["ongoing"] = 2] = "ongoing";
    PlaylistStatus[PlaylistStatus["completed"] = 3] = "completed";
    PlaylistStatus[PlaylistStatus["paused"] = 4] = "paused";
    PlaylistStatus[PlaylistStatus["cancelled"] = 5] = "cancelled";
})(PlaylistStatus || (exports.PlaylistStatus = PlaylistStatus = {}));
var PlaylistType;
(function (PlaylistType) {
    PlaylistType[PlaylistType["video"] = 0] = "video";
    PlaylistType[PlaylistType["image"] = 1] = "image";
    PlaylistType[PlaylistType["text"] = 2] = "text";
})(PlaylistType || (exports.PlaylistType = PlaylistType = {}));
var PlaylistPreviewType;
(function (PlaylistPreviewType) {
    PlaylistPreviewType[PlaylistPreviewType["video"] = 0] = "video";
    PlaylistPreviewType[PlaylistPreviewType["image"] = 1] = "image";
})(PlaylistPreviewType || (exports.PlaylistPreviewType = PlaylistPreviewType = {}));
